File: tutorials/mcp/mcp_dashboards/helpers/astro/host/binding.ts
/**
 * Row shaping for local dashboard hosts (preview + live-local).
 *
 * Mirrors `helpers/python/src/mcp_dashboards/binding.py` so the same
 * `data-unique-attr-href="client_href"` bindings resolve whether rows come
 * from mock JSON or a FastMCP server that already enriched them.
 */
import type { Row } from "./dom.ts";

/** Recursively mirror nested dict/list values as top-level dotted keys. */
export function flattenDottedPaths(value: unknown, prefix = ""): Row {
  const out: Row = {};
  if (value == null || typeof value !== "object" || Array.isArray(value)) return out;

  for (const [key, child] of Object.entries(value as Row)) {
    const path = prefix ? `${prefix}.${key}` : key;
    if (child != null && typeof child === "object" && !Array.isArray(child)) {
      Object.assign(out, flattenDottedPaths(child, path));
    } else if (Array.isArray(child)) {
      child.forEach((item, index) => {
        const itemPath = `${path}.${index}`;
        if (item != null && typeof item === "object" && !Array.isArray(item)) {
          Object.assign(out, flattenDottedPaths(item, itemPath));
        } else {
          out[itemPath] = item;
        }
      });
    } else {
      out[path] = child;
    }
  }
  return out;
}

const ISO_DATE_RE = /^(\d{4}-\d{2}-\d{2})$/;

function dueDateBucket(value: unknown): "none" | "urgent" | "scheduled" {
  if (value == null || value === "") return "none";
  const match = ISO_DATE_RE.exec(String(value).trim());
  if (!match) return "none";
  const iso = match[1]!;
  const now = new Date();
  const month = String(now.getMonth() + 1).padStart(2, "0");
  const day = String(now.getDate()).padStart(2, "0");
  const today = `${now.getFullYear()}-${month}-${day}`;
  return iso <= today ? "urgent" : "scheduled";
}

/** Return a row with nested data, dotted mirrors, and platform attr helpers. */
export function enrichBindingRow(row: Row): Row {
  const flat = flattenDottedPaths(row);
  const merged: Row = { ...row, ...flat };

  const clientId = row.id;
  if (clientId != null) {
    merged.client_href = `#client-${clientId}`;
    merged.client_dom_id = `client-${clientId}`;
  }

  const riskLevel = flat["compliance.risk_level"];
  if (riskLevel != null) {
    merged["compliance.risk_level_tooltip"] = `${riskLevel} risk`;
  }

  merged["case_action.due_bucket"] = dueDateBucket(flat["case_action.due_date"]);

  for (const [key, value] of Object.entries(flat)) {
    if (key.endsWith(".pct") && value != null) {
      merged[`${key}_bar_style`] = `width:${value}%`;
    }
  }

  return merged;
}

File: tutorials/mcp/mcp_dashboards/helpers/python/src/mcp_dashboards/binding.py
"""Row shaping for Unique platform iframe hosts.

The platform resolves ``data-unique-field="identity.name"`` as ``item["identity.name"]``,
not nested ``item.identity.name``. Live-local hosts traverse nested JSON via
``readPath()``. Tool responses therefore carry both shapes: nested domain objects
plus dotted-path mirror keys and a few precomputed attribute values the platform
cannot interpolate from ``{field}`` templates.
"""

from __future__ import annotations

from datetime import date
from typing import Any


def flatten_dotted_paths(value: Any, *, prefix: str = "") -> dict[str, Any]:
    """Recursively mirror nested dict/list values as top-level dotted keys."""
    out: dict[str, Any] = {}
    if isinstance(value, dict):
        for key, child in value.items():
            path = f"{prefix}.{key}" if prefix else str(key)
            if isinstance(child, dict):
                out.update(flatten_dotted_paths(child, prefix=path))
            elif isinstance(child, list):
                for index, item in enumerate(child):
                    item_path = f"{path}.{index}"
                    if isinstance(item, dict):
                        out.update(flatten_dotted_paths(item, prefix=item_path))
                    else:
                        out[item_path] = item
            else:
                out[path] = child
    return out


def due_date_bucket(value: Any, *, today: date | None = None) -> str:
    """Map a due date to ``urgent`` / ``scheduled`` / ``none`` for portfolio filters."""
    if value is None or value == "":
        return "none"
    text = str(value).strip()[:10]
    try:
        due = date.fromisoformat(text)
    except ValueError:
        return "none"
    ref = today or date.today()
    return "urgent" if due <= ref else "scheduled"


def enrich_binding_row(row: dict[str, Any]) -> dict[str, Any]:
    """Return a row dict with nested data, dotted mirrors, and platform attr helpers."""
    flat = flatten_dotted_paths(row)
    merged = {**row, **flat}

    client_id = row.get("id")
    if client_id is not None:
        merged["client_href"] = f"#client-{client_id}"
        merged["client_dom_id"] = f"client-{client_id}"

    risk_level = flat.get("compliance.risk_level")
    if risk_level is not None:
        merged["compliance.risk_level_tooltip"] = f"{risk_level} risk"

    merged["case_action.due_bucket"] = due_date_bucket(flat.get("case_action.due_date"))

    for key, value in flat.items():
        if key.endswith(".pct") and value is not None:
            merged[f"{key}_bar_style"] = f"width:{value}%"

    return merged
